{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "YZyevmLCgYPB"
   },
   "source": [
    "# Exercise 007"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "idtNqXkegYPD",
    "tags": []
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/FAIRChemistry/PythonProgrammingBio24/blob/main/solutions/Exercise007.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Please execute this cell to download the necessary data\n",
    "!wget https://raw.githubusercontent.com/JR-1991/PythonProgrammingBio24/main/scripts/utils.py\n",
    "!wget https://raw.githubusercontent.com/JR-1991/PythonProgrammingBio24/main/data/single_sequence.fasta\n",
    "!pip install pydantic biopython\n",
    "\n",
    "from utils import CODON_TABLE, to_triplets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from Bio import pairwise2\n",
    "\n",
    "def get_identity(seq1: str, seq2: str):\n",
    "    \"\"\"Aligns two sequences using BioPython\n",
    "\n",
    "    Args:\n",
    "        seq1 (str): Query sequence to align to\n",
    "        seq2 (str): Target sequence to align with\n",
    "\n",
    "    Returns:\n",
    "        float: Identity of the resulting alignment\n",
    "\n",
    "    \"\"\"\n",
    "    return pairwise2.align.globalxx(seq1, seq2, score_only=True) / len(seq1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "XbNx_QaHeEcR"
   },
   "source": [
    "## DNASequence class\n",
    "\n",
    "Construct a `DNASequence` class that contains the following attributes:\n",
    "\n",
    "* `id`\n",
    "* `sequence`\n",
    "* `organism`\n",
    "* `gc_content`\n",
    "* `length`\n",
    "* `reverse_complement`\n",
    "\n",
    "Next, implement methods for your class that perform the following tasks:\n",
    "\n",
    "* `to_amino_acid`: Converts the nucelic acid sequence to an amino acid sequence.\n",
    "* `align`: Takes another sequence and aligns it against the instance sequence.\n",
    "* `__repr__`: Define how the contents of your class should be printed.\n",
    "* `from_fasta`: Define a classmethod that parses a single FASTA entry into your class.\n",
    "\n",
    "Demonstrate your class by parsing the `single_sequence.fasta` file either manually or via the `from_fasta`-classmethod.\n",
    "\n",
    "**Tips**\n",
    "\n",
    "> * Feel free to use the `get_identity`-function of the previous exercise.\n",
    "> * When implementing the `classmethod` make sure to check if the format is correct. We have so far followed the `>[Organism]|[ID]\\n[Sequence]` format.\n",
    "> * Translate your sequence using the supported `to_triplets` function and `CODON_TABLE` dictionary.\n",
    "> * Not familiar with reverse complements? Find more info [here](http://genewarrior.com/docs/exp_revcomp.jsp)\n",
    "> * Dont hesitate using the `dataclass` decorator. It can help you in some ways already. Learn more on how to implement `__post_init__` to maximize customizability [here](https://docs.python.org/3/library/dataclasses.html#post-init-processing)\n",
    "> * Python lacks type validation and thus you do have limited control of what flows into your class. [PyDantic](https://docs.pydantic.dev/latest/) is an excellent tool to solve this and other issues. Try it out to make your life easier!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using standard classes\n",
    "\n",
    "The following solution demonstrates an implementation based on Python's native classes. After that a complementary example demonstrates the usage of the library [pyDantic]() which is very practical to solve Python's typing issues and beyond."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "class DNASequence:\n",
    "    \"\"\"\n",
    "    A class to represent a DNA sequence with various utility methods.\n",
    "\n",
    "    Attributes:\n",
    "        sequence (str): The DNA sequence.\n",
    "        id (str): The identifier for the DNA sequence.\n",
    "        organism (str): The organism from which the DNA sequence is derived.\n",
    "        gc_content (float): The GC content of the DNA sequence.\n",
    "        reverse_complement (str): The reverse complement of the DNA sequence.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, sequence: str, id: str, organism: str):\n",
    "        \"\"\"\n",
    "        Initializes the DNASequence with a sequence, id, and organism.\n",
    "\n",
    "        Args:\n",
    "            sequence (str): The DNA sequence.\n",
    "            id (str): The identifier for the DNA sequence.\n",
    "            organism (str): The organism from which the DNA sequence is derived.\n",
    "        \"\"\"\n",
    "        self.sequence = sequence\n",
    "        self.id = id\n",
    "        self.organism = organism\n",
    "        self.gc_content = self.gc_content()\n",
    "        self.reverse_complement = self.reverse_complement()\n",
    "        \n",
    "    def __repr__(self) -> str:\n",
    "        attrs = \"\\n\".join([\n",
    "            f\"\\t{attr}: {value},\"\n",
    "            for attr, value in self.__dict__.items()\n",
    "        ])\n",
    "        \n",
    "        return f\"{type(self).__name__}\\n{attrs}\"\n",
    "\n",
    "        \n",
    "    def gc_content(self) -> float:\n",
    "        \"\"\"\n",
    "        Computes the GC content of the DNA sequence.\n",
    "\n",
    "        Returns:\n",
    "            float: The GC content as a fraction of the total sequence length.\n",
    "        \"\"\"\n",
    "        return (\n",
    "            self.sequence.upper().count(\"G\")\n",
    "            + self.sequence.upper().count(\"C\")\n",
    "        ) / len(self.sequence)\n",
    "    \n",
    "    \n",
    "    def reverse_complement(self) -> str:\n",
    "        \"\"\"\n",
    "        Computes the reverse complement of the DNA sequence.\n",
    "\n",
    "        Returns:\n",
    "            str: The reverse complement of the DNA sequence.\n",
    "        \"\"\"\n",
    "        mapping = {\n",
    "            \"A\": \"T\",\n",
    "            \"T\": \"A\",\n",
    "            \"G\": \"C\",\n",
    "            \"C\": \"G\",\n",
    "        }\n",
    "        \n",
    "        return \"\".join(\n",
    "            [mapping[nuc] for nuc in self.sequence[::-1]]\n",
    "        )\n",
    "    \n",
    "    @classmethod\n",
    "    def from_fasta(cls, fasta: str):\n",
    "        \"\"\"\n",
    "        Creates a DNASequence instance from a FASTA formatted string.\n",
    "\n",
    "        Args:\n",
    "            fasta (str): A string in the format >[Organism]|[ID]\\n[Sequence].\n",
    "\n",
    "        Returns:\n",
    "            DNASequence: An instance of the DNASequence class.\n",
    "\n",
    "        Raises:\n",
    "            ValueError: If the FASTA string is not in the expected format.\n",
    "        \"\"\"\n",
    "        try:\n",
    "            header, seq = fasta.split(\"\\n\")\n",
    "        except ValueError:\n",
    "            raise ValueError(\n",
    "                \"The given DNA sequence does not follow the expected format >[Organism]|[ID]\\n[Sequence]\"\n",
    "            )\n",
    "            \n",
    "        organism, id = header.split(\"|\")\n",
    "        \n",
    "        return cls(\n",
    "            sequence=seq.strip().upper(),\n",
    "            organism=organism.lstrip(\">\"),\n",
    "            id=id,\n",
    "        )\n",
    "    \n",
    "    def to_amino_acid(self) -> str:\n",
    "        \"\"\"\n",
    "        Converts the DNA sequence to its corresponding amino acid sequence.\n",
    "\n",
    "        Returns:\n",
    "            str: The amino acid sequence.\n",
    "        \"\"\"\n",
    "        return \"\".join(\n",
    "            [CODON_TABLE[tripl] for tripl in to_triplets(self.sequence)]\n",
    "        )\n",
    "    \n",
    "    def align(self, seq: 'DNASequence') -> float:\n",
    "        \"\"\"\n",
    "        Aligns the DNA sequence with another sequence and computes the identity.\n",
    "\n",
    "        Args:\n",
    "            seq (DNASequence): Another DNASequence instance to align with.\n",
    "\n",
    "        Returns:\n",
    "            float: The identity score of the alignment.\n",
    "        \"\"\"\n",
    "        if isinstance(seq, DNASequence):\n",
    "            seq = seq.sequence\n",
    "            \n",
    "        return get_identity(self.sequence, seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: 1\n",
      "Organism: ecoli\n",
      "GC: 0.5074971164936563\n",
      "Reverse complement: TTATTTTAAA...\n",
      "Amino acid sequence: MRSRYLLHQY...\n"
     ]
    }
   ],
   "source": [
    "# Load the single sequence\n",
    "fasta = open(\"./single_sequence.fasta\").read().strip()\n",
    "obj = DNASequence.from_fasta(fasta)\n",
    "\n",
    "print(f\"ID: {obj.id}\")\n",
    "print(f\"Organism: {obj.organism}\")\n",
    "print(f\"GC: {obj.gc_content}\")\n",
    "print(f\"Reverse complement: {obj.reverse_complement[:10]}...\")\n",
    "print(f\"Amino acid sequence: {obj.to_amino_acid()[:10]}...\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using dataclasses\n",
    "\n",
    "Dataclasses are a way to simplify the creation of classes that are primarily used to store data. Think of them as a more convenient way to create classes that hold information about something, like a student, book, or car, without having to write a lot of boilerplate code.\n",
    "\n",
    "### Why Use Dataclasses?\n",
    "\n",
    "1. Less Repetition: With dataclasses, you don’t have to manually write the code to store and retrieve the data for each attribute. The dataclass automatically takes care of it.\n",
    "2. Built-In Features: Dataclasses automatically provide useful features like a nice string representation of the object and the ability to compare two objects to see if they are the same.\n",
    "3. Readable Code: Your code becomes cleaner and easier to read because you focus only on what attributes your data should have, not on the repetitive code needed to manage those attributes.\n",
    "\n",
    "### When Dataclasses Shine\n",
    "\n",
    "Dataclasses are particularly useful when you need to:\n",
    "\n",
    "1. Store Simple Data: Whenever you have objects that primarily exist to hold and transfer data.\n",
    "2. Avoid Boilerplate Code: Reduce the amount of repetitive code you have to write, making your codebase cleaner and easier to maintain.\n",
    "3. Compare Objects: Easily compare two objects to see if they are equal, without writing additional code.\n",
    "\n",
    "### Using `__post_init__`\n",
    "\n",
    "Sometimes, you might need to perform some additional actions after the dataclass has been initialized. This is where the __post_init__ method comes in handy. Think of it as a place to do extra setup or validation right after your dataclass has been created. For example, if you need to ensure that a certain attribute always follows a specific rule or needs to be calculated based on other attributes, you can use `__post_init__` to handle that."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from dataclasses import dataclass, field\n",
    "\n",
    "@dataclass\n",
    "class DNASequence:\n",
    "    \"\"\"\n",
    "    A class to represent a DNA sequence with various utility methods.\n",
    "\n",
    "    Attributes:\n",
    "        sequence (str): The DNA sequence.\n",
    "        id (str): The identifier for the DNA sequence.\n",
    "        organism (str): The organism from which the DNA sequence is derived.\n",
    "        gc_content (float): The GC content of the DNA sequence.\n",
    "        reverse_complement (str): The reverse complement of the DNA sequence.\n",
    "    \"\"\"\n",
    "    sequence: str\n",
    "    id: str\n",
    "    organism: str\n",
    "    gc_content: float = field(init=False)\n",
    "    reverse_complement: str = field(init=False)\n",
    "\n",
    "    def __post_init__(self):\n",
    "        \"\"\"\n",
    "        Post-initialization processing to calculate GC content and reverse complement.\n",
    "        \"\"\"\n",
    "        self.gc_content = self.compute_gc_content()\n",
    "        self.reverse_complement = self.compute_reverse_complement()\n",
    "\n",
    "    def compute_gc_content(self) -> float:\n",
    "        \"\"\"\n",
    "        Computes the GC content of the DNA sequence.\n",
    "\n",
    "        Returns:\n",
    "            float: The GC content as a fraction of the total sequence length.\n",
    "        \"\"\"\n",
    "        return (\n",
    "            self.sequence.upper().count(\"G\")\n",
    "            + self.sequence.upper().count(\"C\")\n",
    "        ) / len(self.sequence)\n",
    "    \n",
    "    def compute_reverse_complement(self) -> str:\n",
    "        \"\"\"\n",
    "        Computes the reverse complement of the DNA sequence.\n",
    "\n",
    "        Returns:\n",
    "            str: The reverse complement of the DNA sequence.\n",
    "        \"\"\"\n",
    "        mapping = {\n",
    "            \"A\": \"T\",\n",
    "            \"T\": \"A\",\n",
    "            \"G\": \"C\",\n",
    "            \"C\": \"G\",\n",
    "        }\n",
    "        \n",
    "        return \"\".join(\n",
    "            [mapping[nuc] for nuc in self.sequence[::-1]]\n",
    "        )\n",
    "    \n",
    "    @classmethod\n",
    "    def from_fasta(cls, fasta: str):\n",
    "        \"\"\"\n",
    "        Creates a DNASequence instance from a FASTA formatted string.\n",
    "\n",
    "        Args:\n",
    "            fasta (str): A string in the format >[Organism]|[ID]\\n[Sequence].\n",
    "\n",
    "        Returns:\n",
    "            DNASequence: An instance of the DNASequence class.\n",
    "\n",
    "        Raises:\n",
    "            ValueError: If the FASTA string is not in the expected format.\n",
    "        \"\"\"\n",
    "        try:\n",
    "            header, seq = fasta.split(\"\\n\")\n",
    "        except ValueError:\n",
    "            raise ValueError(\n",
    "                \"The given DNA sequence does not follow the expected format >[Organism]|[ID]\\n[Sequence]\"\n",
    "            )\n",
    "            \n",
    "        organism, id = header.split(\"|\")\n",
    "        \n",
    "        return cls(\n",
    "            sequence=seq.strip().upper(),\n",
    "            organism=organism.lstrip(\">\"),\n",
    "            id=id,\n",
    "        )\n",
    "    \n",
    "    def to_amino_acid(self) -> str:\n",
    "        \"\"\"\n",
    "        Converts the DNA sequence to its corresponding amino acid sequence.\n",
    "\n",
    "        Returns:\n",
    "            str: The amino acid sequence.\n",
    "        \"\"\"\n",
    "        return \"\".join(\n",
    "            [CODON_TABLE[tripl] for tripl in to_triplets(self.sequence)]\n",
    "        )\n",
    "    \n",
    "    def align(self, seq: 'DNASequence') -> float:\n",
    "        \"\"\"\n",
    "        Aligns the DNA sequence with another sequence and computes the identity.\n",
    "\n",
    "        Args:\n",
    "            seq (DNASequence): Another DNASequence instance to align with.\n",
    "\n",
    "        Returns:\n",
    "            float: The identity score of the alignment.\n",
    "        \"\"\"\n",
    "        if isinstance(seq, DNASequence):\n",
    "            seq = seq.sequence\n",
    "            \n",
    "        return get_identity(self.sequence, seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: 1\n",
      "Organism: ecoli\n",
      "GC: 0.5074971164936563\n",
      "Reverse complement: TTATTTTAAA...\n",
      "Amino acid sequence: MRSRYLLHQY...\n"
     ]
    }
   ],
   "source": [
    "# Load the single sequence\n",
    "fasta = open(\"./single_sequence.fasta\").read().strip()\n",
    "obj = DNASequence.from_fasta(fasta)\n",
    "\n",
    "print(f\"ID: {obj.id}\")\n",
    "print(f\"Organism: {obj.organism}\")\n",
    "print(f\"GC: {obj.gc_content}\")\n",
    "print(f\"Reverse complement: {obj.reverse_complement[:10]}...\")\n",
    "print(f\"Amino acid sequence: {obj.to_amino_acid()[:10]}...\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PyDantic\n",
    "\n",
    "[Pydantic](https://docs.pydantic.dev/latest/) is a library that helps you define data models with built-in data validation. Think of it as a way to create classes that not only hold information but also automatically check that the information is correct. For instance, if you expect a user’s age to be a positive integer, Pydantic ensures it is, right when you create the object.\n",
    "\n",
    "### Why Use Pydantic?\n",
    "\n",
    "1. Automatic Validation: Pydantic checks the data types and values for you, ensuring that the data your program works with is always valid.\n",
    "2. Easy to Use: It’s straightforward to define data models with Pydantic, making your code cleaner and easier to understand.\n",
    "3. Error Handling: Pydantic provides clear and helpful error messages if the data doesn’t meet the expected format, making debugging easier.\n",
    "4. Built-In Parsing: Pydantic can automatically convert input data to the correct type, saving you from writing additional conversion code.\n",
    "\n",
    "### When Pydantic Shines\n",
    "\n",
    "Pydantic is especially useful when you need to:\n",
    "\n",
    "1. Validate Input Data: Ensure that the data coming into your program (from user input, APIs, etc.) is in the correct format and meets all your requirements.\n",
    "2. Simplify Data Handling: Reduce the amount of manual validation and conversion code you need to write, making your codebase more maintainable.\n",
    "3. Work with APIs: When dealing with data from external sources, Pydantic helps ensure that the data you receive and send out is always in the expected format.\n",
    "\n",
    "Pydantic makes your code more reliable by automatically checking and validating the data, allowing you to focus on building features instead of writing boilerplate validation code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "id": "yubzZtmteEcR",
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pydantic import BaseModel, computed_field\n",
    "\n",
    "class DNASequence(BaseModel):\n",
    "    \"\"\"\n",
    "    A class to represent a DNA sequence with various utility methods.\n",
    "\n",
    "    Attributes:\n",
    "        sequence (str): The DNA sequence.\n",
    "        id (str): The identifier for the DNA sequence.\n",
    "        organism (str): The organism from which the DNA sequence is derived.\n",
    "    \"\"\"\n",
    "\n",
    "    sequence: str\n",
    "    id: str\n",
    "    organism: str\n",
    "    \n",
    "    @computed_field\n",
    "    @property\n",
    "    def gc_content(self) -> float:\n",
    "        \"\"\"\n",
    "        Computes the GC content of the DNA sequence.\n",
    "\n",
    "        Returns:\n",
    "            float: The GC content as a fraction of the total sequence length.\n",
    "        \"\"\"\n",
    "        return (\n",
    "            self.sequence.upper().count(\"G\")\n",
    "            + self.sequence.upper().count(\"C\")\n",
    "        ) / len(self.sequence)\n",
    "    \n",
    "    @computed_field\n",
    "    @property\n",
    "    def reverse_complement(self) -> str:\n",
    "        \"\"\"\n",
    "        Computes the reverse complement of the DNA sequence.\n",
    "\n",
    "        Returns:\n",
    "            str: The reverse complement of the DNA sequence.\n",
    "        \"\"\"\n",
    "        mapping = {\n",
    "            \"A\": \"T\",\n",
    "            \"T\": \"A\",\n",
    "            \"G\": \"C\",\n",
    "            \"C\": \"G\",\n",
    "        }\n",
    "        \n",
    "        return \"\".join(\n",
    "            [mapping[nuc] for nuc in self.sequence[::-1]]\n",
    "        )\n",
    "    \n",
    "    @classmethod\n",
    "    def from_fasta(cls, fasta: str):\n",
    "        \"\"\"\n",
    "        Creates a DNASequence instance from a FASTA formatted string.\n",
    "\n",
    "        Args:\n",
    "            fasta (str): A string in the format >[Organism]|[ID]\\n[Sequence].\n",
    "\n",
    "        Returns:\n",
    "            DNASequence: An instance of the DNASequence class.\n",
    "\n",
    "        Raises:\n",
    "            ValueError: If the FASTA string is not in the expected format.\n",
    "        \"\"\"\n",
    "        try:\n",
    "            header, seq = fasta.split(\"\\n\")\n",
    "        except ValueError:\n",
    "            raise ValueError(\n",
    "                \"The given DNA sequence does not follow the expected format >[Organism]|[ID]\\n[Sequence]\"\n",
    "            )\n",
    "            \n",
    "        organism, id = header.split(\"|\")\n",
    "        \n",
    "        return cls(\n",
    "            sequence=seq.strip().upper(),\n",
    "            organism=organism.lstrip(\">\"),\n",
    "            id=id,\n",
    "        )\n",
    "    \n",
    "    def to_amino_acid(self) -> str:\n",
    "        \"\"\"\n",
    "        Converts the DNA sequence to its corresponding amino acid sequence.\n",
    "\n",
    "        Returns:\n",
    "            str: The amino acid sequence.\n",
    "        \"\"\"\n",
    "        return \"\".join(\n",
    "            [CODON_TABLE[tripl] for tripl in to_triplets(self.sequence)]\n",
    "        )\n",
    "    \n",
    "    def align(self, seq: 'DNASequence') -> float:\n",
    "        \"\"\"\n",
    "        Aligns the DNA sequence with another sequence and computes the identity.\n",
    "\n",
    "        Args:\n",
    "            seq (DNASequence): Another DNASequence instance to align with.\n",
    "\n",
    "        Returns:\n",
    "            float: The identity score of the alignment.\n",
    "        \"\"\"\n",
    "        if isinstance(seq, DNASequence):\n",
    "            seq = seq.sequence\n",
    "            \n",
    "        return get_identity(self.sequence, seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: 1\n",
      "Organism: ecoli\n",
      "GC: 0.5074971164936563\n",
      "Reverse complement: TTATTTTAAA...\n",
      "Amino acid sequence: MRSRYLLHQY...\n"
     ]
    }
   ],
   "source": [
    "# Load the single sequence\n",
    "fasta = open(\"./single_sequence.fasta\").read().strip()\n",
    "obj = DNASequence.from_fasta(fasta)\n",
    "\n",
    "print(f\"ID: {obj.id}\")\n",
    "print(f\"Organism: {obj.organism}\")\n",
    "print(f\"GC: {obj.gc_content}\")\n",
    "print(f\"Reverse complement: {obj.reverse_complement[:10]}...\")\n",
    "print(f\"Amino acid sequence: {obj.to_amino_acid()[:10]}...\")"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
